### PR TITLE
[Feature] Wallet - Handle blockchain status

### LINKED ISSUE #18826: Wallet - Handle blockchain status signal
# Feature Issue

`status-go` sends a `wallet-blockchain-status-changed` signal with chain data (whether it's up or down).

If any of the chains is down (due to provider, network connectivity, or API exhaustion), we need to show a toast* to the user.

### User Story

> As a user, I want to know if any chain is down so that I will not perform any activity on it. 

### Description

On receiving wallet event *`wallet-blockchain-status-changed`*, the mobile app should check whether any chains are down. If there is any, we should inform the user with a toast.

### Acceptance Criteria

- Toast is shown if any of the chains are down

### Notes

To simulate chain down for testing/integration, use the wrong RPC URL in `status-go`.

*The design team is working on a sticky banner. For now, we will use Toast to inform the user

### PR DESCRIPTION
**Pull request by smohamedjavid**
fixes #18826 

### Summary

This PR adds the feature to handle blockchain status signals from the `status-go` and shows a toast if the provider for any of the blockchains is down.

Additionally, this PR refactors a few areas of the wallet.

### Testing notes

To simulate the chains down, we need to create a test branch in status-go with completely wrong network RPC URLs and use it. I will create it when we start testing.

A regression test on the wallet is appreciated as this PR refactors a few areas of the Wallet.

### Platforms

- Android
- iOS

### Steps to test

- Open Status
- Login to your profile
- Verify that toast is shown if any of the blockchains is down

status: ready

### CODE REVIEW

File: src/status_im/constants.cljs
Comment (smohamedjavid) [inline src/status_im/constants.cljs]: Just renamed the chain-id names to a bit more specific one.

File: src/status_im/contexts/wallet/data_store.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/data_store.cljs:63]: We have been unnecessarily sending huge amounts of tokens and collectible data whenever we edit an account.

File: src/status_im/contexts/wallet/signals.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/signals.cljs:1]: Cleaned up a bit

File: src/status_im/contexts/wallet/common/utils.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/common/utils.cljs]: Added the Sepolia check.

File: src/status_im/contexts/wallet/signals.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/signals.cljs:1]: Cleaned up a bit

File: src/status_im/contexts/wallet/events.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/events.cljs]: To prevent showing the toast if a testnet chain is down and the user is not in testnet mode.

File: src/status_im/constants.cljs
Comment (OmarBasem) [inline src/status_im/constants.cljs]: what about calling the test networks `ethereum-testnet`, `optimism-testnet`, `arbitrum-testnet` ?

**Comment by VolodLytvynenko** [comment]
Hey @smohamedjavid thank you for PR.  Here is the result. Looks like the problem is not provider related

### Actual result:
The routes are not found and the toast is not shown

https://github.com/status-im/status-mobile/assets/52490791/4b88dcdb-6f74-441a-8e09-4d15db4322c5


### Expected result:
The toast is shown if the provider for any of the blockchains is down


### Logs:
[logs (5).zip](https://github.com/status-im/status-mobile/files/14309058/logs.5.zip)

**Comment by VolodLytvynenko** [comment]
I noticed [this](https://github.com/status-im/status-mobile/pull/18850#issuecomment-1948041140) issue is reproducible for me only when I am at home wifi. Now, I am in the coworking space, and it is not reproducible for me anymore (now routes are successfully found). I thought it might be a low network issue, but that's strange. I tried to emulate a low network connection using Andoird Studio or mobile network on real device, and the routes were successfully found for me. I will try to reproduce this at home one more time and take the logcat logs

**Comment by VolodLytvynenko** [comment]
@smohamedjavid just got this issue in the coworking space. Here are the logcat logs:
[logs2.txt](https://github.com/status-im/status-mobile/files/14309525/logs2.txt)

File: src/status_im/constants.cljs
Comment (smohamedjavid) [reply src/status_im/constants.cljs]: I would love to. The reason why we renamed it specifically is that we have two test networks (Goerli and Sepolia) at this moment. In future, One will supersede the old, and another will be in the pipeline for testing/integration (It's a cycle).  Additionally, this rename goes well with status-go.

**Comment by smohamedjavid** [comment]
@VolodLytvynenko - Thank you for debugging the issue. Duly appreciate it! 
I will check the logs and will update my findings in a while.

**Review by J-Son89 — APPROVED** [review]

File: src/status_im/contexts/wallet/data_store.cljs
Comment (ulisesmac) [reply src/status_im/contexts/wallet/data_store.cljs:63]: Nice!
I've faced some performance issues while just updating the profile!

**Comment by smohamedjavid** [comment]
@VolodLytvynenko - The fix is on https://github.com/status-im/status-mobile/pull/18879 for grouping any bugs caused by the PR changes (if any).

File: src/status_im/contexts/wallet/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/events.cljs]: This `filter-fn` could be called `disabled-chain-id?` (or another better name, just thinking that `(remove filter-fn coll)` is unclear

File: src/status_im/contexts/wallet/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/events.cljs]: I think that `set` is not needed, actually, sets have no order, so this might lead to an inconsistent resulting string

File: src/status_im/contexts/wallet/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/events.cljs]: We can extract `:fx` to make the code easier to understand. `:fx` is able to handle `nil`:

```clojure
{:db (assoc-in db [:wallet :statuses :blockchains] chains)
 :fx [(when chains-down?
        [:dispatch
         [:toasts/upsert
          {:id       :chains-down
           :type     :negative
           :text     (i18n/label :t/provider-is-down {:chains chain-names})
           :duration 10000}]])]}
```

Also probably we don't need to keep the vector outside `when`, so you could try:

```clojure
{:db ...
 :fx (when chains-down?
       [[:dispatch ...]])}
```

**Review by ulisesmac — APPROVED** [review]
Hey @smohamedjavid 

Sorry for the late review, just added some small comments :+1:

**Review by OmarBasem — APPROVED** [review]

File: src/status_im/contexts/wallet/events.cljs
Comment (smohamedjavid) [reply src/status_im/contexts/wallet/events.cljs]: Agree 👍 I will update it

File: src/status_im/contexts/wallet/events.cljs
Comment (smohamedjavid) [reply src/status_im/contexts/wallet/events.cljs]: When the Sepolia test net is enabled, we can see the Goerli status too. We don't have to show the chain name (Eg: `Ethereum`) twice. So, I used set to get the distinct items.

File: src/status_im/contexts/wallet/events.cljs
Comment (smohamedjavid) [reply src/status_im/contexts/wallet/events.cljs]: Yes @ulisesmac, It's just I thought why pass nil when we can make use of `cond->`.  We don't have a lot of conditions here.

File: src/status_im/contexts/wallet/events.cljs
Comment (ulisesmac) [reply src/status_im/contexts/wallet/events.cljs]: There's a `distinct` clojure function :)

Comment (smohamedjavid) [reply src/status_im/contexts/wallet/events.cljs]: Ah yes! 😄 Totally forgot that. I will update. 👍  Thanks!

File: src/status_im/contexts/wallet/events.cljs
Comment (ulisesmac) [reply src/status_im/contexts/wallet/events.cljs]: The "new" re-frame API (a vector of `:fx` able to handle `nil`) was thought to give more flexibility.

I think adding or removing `nil` doesn't change anything.


I was suggesting this because IMO it's easier to "parse"  in developer's mind, since we'd be using simple patterns, I'm not saying `cond->` is bad itself, but I just think it's not needed.

Anyway, this might be a matter of prefernces, so it's ok if you apply it or not :+1:

Comment (smohamedjavid) [reply src/status_im/contexts/wallet/events.cljs]: I agree it's easy to read the code. I'm not saying the patterns we used to are wrong. Apologies, if you felt like that 🙏. 

It's just my experiences with some events in the early stages which made me use `cond->` to resolve a bug (encountered due to `nil`). But that was before we upgraded the re-frame and `:fx` is life saver as it can handle `nil` and doesn't throw any error. I will update it 👍

**Comment by VolodLytvynenko** [comment]
hey @smohamedjavid could you please rebase and resolve conflicts for this PR? thanx!

**Comment by smohamedjavid** [comment]
> hey @smohamedjavid could you please rebase and resolve conflicts for this PR? thanx!

Hey @VolodLytvynenko - If it's not too much trouble, can we deprioritise testing this PR now? 
We need to fast-track this PR: https://github.com/status-im/status-mobile/pull/18917. 

We will return to this PR after that as I need to help you in testing as well (creating a temporary status-go to simulate the provider down cases if required).

**Comment by VolodLytvynenko** [comment]
> > hey @smohamedjavid could you please rebase and resolve conflicts for this PR? thanx!
> 
> Hey @VolodLytvynenko - If it's not too much trouble, can we deprioritise testing this PR now? We need to fast-track this PR: #18917.
> 
> We will return to this PR after that as I need to help you in testing as well (creating a temporary status-go to simulate the provider down cases if required).

Hey @smohamedjavid Sure. Thank you!

**Comment by smohamedjavid** [comment]
@VolodLytvynenko - This PR is ready for testing

**Comment by qoqobolo** [comment]
Hey @smohamedjavid, could you please rebase this PR and create a status-go test branch to simulate the chains down?

**Comment by smohamedjavid** [comment]
@qoqobolo - In the builds: https://github.com/status-im/status-mobile/pull/18850#issuecomment-1947378104 

Build #7 is WITHOUT any changes in `status-go` (just a rebase with the latest develop - d637f69a450d9ee16072d75033b5f75e47f58ef1).

Build #8 is with changes in `status-go` to simulate chain down.

Let me know if you need anything. Thanks!

**Comment by qoqobolo** [comment]
> @qoqobolo - In the builds: https://github.com/status-im/status-mobile/pull/18850#issuecomment-1947378104
>
> Build https://github.com/status-im/status-mobile/issues/7 is WITHOUT any changes in status-go (just a rebase with the latest develop - https://github.com/status-im/status-mobile/commit/d637f69a450d9ee16072d75033b5f75e47f58ef1).
>
> Build https://github.com/status-im/status-mobile/pull/8 is with changes in status-go to simulate chain down.

Thanks @smohamedjavid! PR looks good to me, but I have a question

When I check build #7, where the chains are not supposed to be down, I still see the banner when entering the amount of the asset being sent. 
And depending on the amount, the banner says that the chains. 

Screenshot 2024-03-01 at 13 24 55

It seems the banner appears when the user does not have enough ETH on other networks (?)

Could you check and confirm if it's not expected for that build, please?





For example, this user only has ETH on Arbitrium:

Screenshot 2024-03-01 at 15 11 35

More examples when user has some ETH on Mainnet, Optimis and Arbitrium:

Screenshot 2024-03-01 at 15 13 54

Screenshot 2024-03-01 at 15 14 06

Screenshot 2024-03-01 at 15 14 17

**Comment by smohamedjavid** [comment]
> Thanks @smohamedjavid! PR looks good to me, but I have a question
> 
> When I check build #7, where the chains are not supposed to be down, I still see the banner when entering the amount of the asset being sent. And depending on the amount, the banner says that the chains.
> It seems the banner appears when the user does not have enough ETH on other networks (?)
> 
> Could you check and confirm if it's not expected for that build, please?
> 
> For example, this user only has ETH on Arbitrium:
> 
> More examples when user has some ETH on Mainnet, Optimis and Arbitrium:

Checking it with the Desktop team. It should not happen. 🙏

**Comment by smohamedjavid** [comment]
@qoqobolo - You are right. It's due to insufficient gas fees (due to no balance) on those chains that make false positive results.

There is a fix merged into status-go https://github.com/status-im/status-go/commit/004bd4e79a783a32f5238a80ea8d490275596180 by the desktop team. 

I updated this PR to use that version tag. One known issue in this PR is the Mainnet Sepolia balances are not fetched due to POKT dropping support for it. If that's not a blocker for testing, Please proceed with verifying the fix in this PR.

If not, there is a separate PR (https://github.com/status-im/status-mobile/pull/19040) for removing Goerli RPC and I have included the fix for Sepolia as well (including the latest commits from status-go and the existing insufficient gas fee fix). We can wait for that to be merged.

My suggestion is to keep this PR on hold until the above mentioned PR (#19040) is merged (to capture and fix any bugs due to status-go changes) and keep this PR from any status-go changes.

**Comment by qoqobolo** [comment]
@smohamedjavid gotcha, thank you!
I see that https://github.com/status-im/status-mobile/pull/19040 is already merged and you've rebased the branch, so I'll just check this PR soon 👍

**Comment by qoqobolo** [comment]
@smohamedjavid don't see any problems anymore. 
But the only thing - I did not check the simulated situation when the chain was down this time. 
If you are sure that nothing has changed with the last commits, and there is no need for another check, then the PR can be merged.

**Comment by smohamedjavid** [comment]
> @smohamedjavid don't see any problems anymore. But the only thing - I did not check the simulated situation when the chain was down this time. If you are sure that nothing has changed with the last commits, and there is no need for another check, then the PR can be merged.

@qoqobolo - Thank you for testing this PR. This PR (with the latest commit/update - https://github.com/status-im/status-mobile/pull/18850/commits) doesn't update any status-go version. It's all using the existing functionalities/bug fixes from status-go now.

> I did not check the simulated situation when the chain was down this time.

I can push a temp commit to point to the temp status-go branch. But, it will be the same as Build #8. Nothing new to test. 

We can merge this PR 👍

**Pull request merged**
